parlai/core/torch_agent.py

```diff
@@ -0,0 +1,260 @@
+
+        Returns a namedtuple Batch that contains, in order, batched text
+        vectors, batched label vectors, batched labels, indices of valid
+        observations, and lengths of label vectors.
+
+        :param obs_batch: list of vectorized observations
+        :param sort: default True, orders the observations by length of vector
+        :param is_valid: default lambda, determines if an observation is valid
+        """
+        if len(obs_batch) == 0:
+            return Batch()
+
+        valid_obs = [(i, ex) for i, ex in enumerate(obs_batch) if is_valid(ex)]
+
+        if len(valid_obs) == 0:
+            return Batch()
+
+        valid_inds, exs = zip(*valid_obs)
+
+        x_text = [ex['text_vec'] for ex in exs]
+        x_lens = [ex.shape[0] for ex in x_text]
+
+        if sort:
+            ind_sorted = sorted(range(len(x_lens)), key=lambda k: -x_lens[k])
+
+            exs = [exs[k] for k in ind_sorted]
+            valid_inds = [valid_inds[k] for k in ind_sorted]
+            x_text = [x_text[k] for k in ind_sorted]
+
+        padded_xs = torch.LongTensor(len(exs),
+                                     max(x_lens)).fill_(self.NULL_IDX)
+        if self.use_cuda:
+            padded_xs = padded_xs.cuda()
+
+        for i, ex in enumerate(x_text):
+            padded_xs[i, :ex.shape[0]] = ex
+
+        xs = padded_xs
+
+        eval_labels_avail = any(['eval_labels_vec' in ex for ex in exs])
+        labels_avail = any(['labels_vec' in ex for ex in exs])
+        some_labels_avail = eval_labels_avail or labels_avail
+
+        # set up the target tensors
+        ys = None
+        labels = None
+        if some_labels_avail:
+            # randomly select one of the labels to update on (if multiple)
+            if labels_avail:
+                num_choices = [len(ex.get('labels_vec', [])) for ex in exs]
+                choices = [random.choice(range(num)) if num != 0 else -1
+                           for num in num_choices]
+                label_vecs = [ex['labels_vec'][choices[i]]
+                              if choices[i] != -1 else torch.LongTensor([])
+                              for i, ex in enumerate(exs)]
+                labels = [ex['labels'][choices[i]]
+                          if choices[i] != -1 else ''
+                          for i, ex in enumerate(exs)]
+            else:
+                num_choices = [len(ex.get('eval_labels_vec', [])) for ex in exs]
+                choices = [random.choice(range(num)) if num != 0 else -1
+                           for num in num_choices]
+                label_vecs = [ex['eval_labels_vec'][choices[i]]
+                              if choices[i] != -1 else torch.LongTensor([])
+                              for i, ex in enumerate(exs)]
+                labels = [ex['eval_labels'][choices[i]]
+                          if choices[i] != -1 else ''
+                          for i, ex in enumerate(exs)]
```

Comment:
should combine this repeated code unless we move eval_labels completely (@jaseweston and I had discussed a way of doing this but it's not in the works yet)